### PR TITLE
[Feature] temporary table(part-0): introduce TemporaryTableMgr

### PR DESCRIPTION
**Pull request by silverbullet233**
## What I'm doing:

the first part of work split from #43048 .

introduce TemporaryTableMgr for temporary tables,  I will implement temporary table features based on these data structures.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/server/Session.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                    .collect(Collectors.toMap(Map.Entry::getKey,
+                            e -> new ConcurrentHashMap<>(e.getValue())));
+        }
+        return tempCopy;
+    }
+
+    public List<String> listTemporaryTables(long databaseId) {
+        List<String> tableNames = Lists.newArrayList();
+        if (temporaryTables.containsKey(databaseId)) {
+            synchronized (temporaryTables) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/server/Session.java:89]: The if should be protected by this synchronized.

File: fe/fe-core/src/main/java/com/starrocks/server/Session.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+// used for temporary tables to store all temporary table ids created by a session,
+// maintain the mapping relationship of databaseId -> tableName -> tableId.
+public class Session {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/server/Session.java:32]: you should add a UT for Session class too.

File: fe/fe-core/src/main/java/com/starrocks/server/Session.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    private static final Logger LOG = LogManager.getLogger(Session.class);
+    private UUID id;
+
+    private Map<Long, Map<String, Long>> temporaryTables;
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/server/Session.java:36]: give some comment for this map. Also you can use Table directly.

File: fe/fe-core/src/main/java/com/starrocks/server/Session.java (hunk 2 above)
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/server/Session.java:32]: Session may be not a good class name now, it's looks only used on temporary table, but many feature(user variable, session variable) is session level, but doesn't implement in here, I think it's confusing.

of course, we can refactor code and move feature to here, but I think `TemporaryTableMgr` is more like `SessionMgr`?

Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/server/Session.java:32]: > Session may be not a good class name now, it's looks only used on temporary table, but many feature(user variable, session variable) is session level, but doesn't implement in here, I think it's confusing.
> 
> of course, we can refactor code and move feature to here, but I think `TemporaryTableMgr` is more like `SessionMgr`?

any suggestion about naming? @Seaven

File: fe/fe-core/src/main/java/com/starrocks/server/Session.java (hunk 3 above)
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/server/Session.java:36]: > give some comment for this map. Also you can use Table directly.

I have considered it, but in the current metadata, it seems that Table is only managed by Database. Do we need to respect this design? If not needed, I think we can use Table here directly. @imay

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/server/TemporaryTableMgr.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        return tablesMap.keySet();
+    }
+
+
+    @VisibleForTesting
+    public void clear() {
+        if (tablesMap != null) {
+            tablesMap.clear();
+        }
+        System.gc();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/server/TemporaryTableMgr.java:167]: Is this necessary? And can this method be called in production env or just for testing purpose?

Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/server/TemporaryTableMgr.java:167]: > Is this necessary? And can this method be called in production env or just for testing purpose?

just for test， won't be used in production env @nshangyiming

**Review by nshangyiming — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by silverbullet233** [comment]
@mergify backport branch-3.3

**Pull request merged**